tools/release-ci/pkg/buildrunner/runbuild.go

```diff
@@ -0,0 +1,211 @@
+	}
+	return result[1], nil
+}
+
+func extractExistingBuildName(str string) (string, error) {
+	existingBuildRegex := regexp.MustCompile("AlreadyExists.*builds \"([^\"]*)\" already exists")
+	result := existingBuildRegex.FindStringSubmatch(str)
+	if len(result) != 2 {
+		return "", fmt.Errorf("Unexpected result: %s", str)
+	}
+	return result[1], nil
+}
+
+func followBuild(name string) error {
+	err := waitForRunning(name)
+	if err != nil {
+		return err
+	}
+	err = logBuild(name)
+	if err != nil {
+		return err
+	}
+	return checkBuildSuccess(name)
+}
+
+func isNewOrPending(status string) bool {
+	return status == "New" || status == "Pending"
+}
+
+func isComplete(status string) bool {
+	return status != "New" && status != "Pending" && status != "Running"
+}
+
+func isSuccessful(status string) bool {
+	return status == "Complete"
+}
+
+func getBuildStatus(name string) (string, error) {
+	result, err := runCmd("oc", "get", "build", name, "-o", "jsonpath={ .status.phase }")
+	if err != nil {
+		return "", fmt.Errorf("error getting build status for %s: %v, output: %s", name, err, result)
+	}
+	return strings.TrimSpace(result), nil
+}
+
+func getBuildCreationTime(name string) (time.Time, error) {
+	result, err := runCmd("oc", "get", "build", name, "-o", "jsonpath={ .metadata.creationTimestamp }")
+	if err != nil {
+		return time.Time{}, fmt.Errorf("error getting build creation timestamp for %s: %v, output: %s", name, err, result)
+	}
+	ts, err := time.Parse(time.RFC3339, strings.TrimSpace(result))
+	if err != nil {
+		return time.Time{}, fmt.Errorf("error parsing build creation timestamp for %s: %v, input: %s", name, err, result)
+	}
+	return ts, nil
+}
+
+func waitForRunning(name string) error {
+	status, err := getBuildStatus(name)
+	if err != nil {
+		return err
+	}
+	if !isNewOrPending(status) {
+		return nil
+	}
+	// Begin wait
+	creationTime, err := getBuildCreationTime(name)
+	if err != nil {
+		return err
+	}
+	for {
+		status, err := getBuildStatus(name)
+		if err != nil {
+			return err
+		}
+		if !isNewOrPending(status) {
+			return nil
+		}
+		if time.Since(creationTime) > WaitForRunningTimeout {
+			return fmt.Errorf("Timed out waiting for build to start")
+		}
+		time.Sleep(5)
+	}
+	return nil
+}
+
+func logBuild(name string) error {
+	status, err := getBuildStatus(name)
+	if err != nil {
+		return err
+	}
+	if status == "Error" {
+		fmt.Printf("Build is in Error state. Cannot print log")
+		return nil
+	}
+	return printBuildLog(name, status == "Running")
+}
+
+func printBuildLog(name string, follow bool) error {
+	var args []string
+	if follow {
+		args = []string{"logs", "-f", fmt.Sprintf("build/%s", name)}
+	} else {
+		args = []string{"logs", fmt.Sprintf("build/%s", name)}
+	}
+
+	cmd := exec.Command("oc", args...)
+	cmd.Stdout = os.Stdout
+	cmd.Stderr = os.Stderr
+	return cmd.Run()
+}
+
+func checkBuildSuccess(name string) error {
+	status, err := getBuildStatus(name)
+	if !isComplete(status) {
```

Comment:
isn't this a logic error?

Reply:
based on when we call this func

Reply:
No, following the build doesn't guarantee that the build will be done running. I've made that wrong assumption before in pipelines. There's all kinds of reasons why the `logs -f` command can terminate early and I really want to make sure that the build is done.

Reply:
ack